src/platforms/macos/daemon/dnsutilsmacos.cpp

```diff
@@ -19,205 +17,52 @@ Logger logger("DnsUtilsMacos");
-        kCFAllocatorSystemDefault, nullptr,
-        CFSTR("Setup:/Network/Service/%s/DNS"), qPrintable(uuid));
-    if (!dnsPath) {
-      continue;
-    }
-    SCDynamicStoreSetValue(m_scStore, dnsPath, dnsConfig);
-    CFRelease(dnsPath);
-  }
-  return true;
+  // Launch the DNS manager process.
+  m_dnsManager.start(qApp->applicationFilePath(), args);
+  return m_dnsManager.waitForStarted();
 }
 
 bool DnsUtilsMacos::restoreResolvers() {
-  for (const QString& uuid : m_prevServices.keys()) {
-    CFStringRef path = CFStringCreateWithFormat(
-        kCFAllocatorSystemDefault, nullptr,
-        CFSTR("Setup:/Network/Service/%s/DNS"), qPrintable(uuid));
-
-    logger.debug() << "Restoring DNS config for" << uuid;
-    const DnsBackup& backup = m_prevServices[uuid];
-    if (backup.isValid()) {
-      CFMutableDictionaryRef config;
-      config = CFDictionaryCreateMutable(kCFAllocatorSystemDefault, 0,
-                                         &kCFCopyStringDictionaryKeyCallBacks,
-                                         &kCFTypeDictionaryValueCallBacks);
-
-      cfDictSetString(config, kSCPropNetDNSDomainName, backup.m_domain);
-      cfDictSetStringList(config, kSCPropNetDNSSearchDomains, backup.m_search);
-      cfDictSetStringList(config, kSCPropNetDNSServerAddresses,
-                          backup.m_servers);
-      cfDictSetStringList(config, kSCPropNetDNSSortList, backup.m_sortlist);
-      SCDynamicStoreSetValue(m_scStore, path, config);
-      CFRelease(config);
-    } else {
-      SCDynamicStoreRemoveValue(m_scStore, path);
-    }
-    CFRelease(path);
-  }
-
-  m_prevServices.clear();
-  return true;
-}
-
-void DnsUtilsMacos::backupService(const QString& uuid) {
-  DnsBackup backup;
-  CFStringRef path = CFStringCreateWithFormat(
-      kCFAllocatorSystemDefault, nullptr,
-      CFSTR("Setup:/Network/Service/%s/DNS"), qPrintable(uuid));
-  CFDictionaryRef config =
-      (CFDictionaryRef)SCDynamicStoreCopyValue(m_scStore, path);
-  auto serviceGuard = qScopeGuard([&] {
-    if (config) {
-      CFRelease(config);
-    }
-    CFRelease(path);
-  });
-
-  // Parse the DNS protocol entry and save it for later.
-  if (config) {
-    CFTypeRef value;
-    value = CFDictionaryGetValue(config, kSCPropNetDNSDomainName);
-    if (value) {
-      backup.m_domain = cfParseString(value);
-    }
-    value = CFDictionaryGetValue(config, kSCPropNetDNSServerAddresses);
-    if (value) {
-      backup.m_servers = cfParseStringList(value);
-    }
-    value = CFDictionaryGetValue(config, kSCPropNetDNSSearchDomains);
-    if (value) {
-      backup.m_search = cfParseStringList(value);
-    }
-    value = CFDictionaryGetValue(config, kSCPropNetDNSSortList);
-    if (value) {
-      backup.m_sortlist = cfParseStringList(value);
-    }
+  if (m_dnsManager.state() == QProcess::NotRunning) {
+    return true;
   }
 
-  m_prevServices[uuid] = backup;
+  m_dnsManager.terminate();
+  return m_dnsManager.waitForFinished();
```

Comment:
Should an explicit timeout be used here or is the default OK? Looks like the default is 30s, which may be too long.

Reply:
The timeout here would only catch the case where the `macosdnsmanager` command hangs or something goes horribly wrong, so I think the default is fine here.